<!DOCTYPE html><html lang="en-us" >

<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  
  
  
  <meta name="generator" content="Wowchemy 5.0.0-beta.0 for Hugo">
  

  

  
  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Chen-Zhu Wang">

  
  
  
    
  
  <meta name="description" content="无参转录组分析的小例子.">

  
  <link rel="alternate" hreflang="en-us" href="www.moths.life/en/project/bioinformatics/">

  







  




  
  
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  

  
  
  
  <meta name="theme-color" content="#3f51b5">
  

  
  

  
  
  
  
    
    
      <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.9.0/css/academicons.min.css" integrity="sha512-W4yqoT1+8NLkinBLBZko+dFB2ZbHsYLDdr50VElllRcNt2Q4/GSs6u71UHKxB7S6JEMCp5Ve4xjh3eGQl/HRvg==" crossorigin="anonymous">
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css" integrity="sha256-FMvZuGapsJLjouA6k7Eo2lusoAX9i0ShlWFG6qt7SLc=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.css" integrity="sha256-Vzbj7sDDS/woiFS3uNKo8eIuni59rjyNGtXfstRzStA=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-light">
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/dracula.min.css" crossorigin="anonymous" title="hl-dark" disabled>
        
      
    

    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.7.1/leaflet.min.css" integrity="sha512-1xoFisiGdy9nvho8EgXuXvnpR5GAMSjFwp40gSRE3NwdUdIMIKuPa7bqoUhLD0O/5tPNhteAsE5XyyMi5reQVA==" crossorigin="anonymous">
    

    

    
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
    
      

      
      

      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/lazysizes/5.2.2/lazysizes.min.js" integrity="sha512-TmDwFLhg3UA4ZG0Eb4MIyT1O1Mb+Oww5kFG0uHqXsdbyZz9DcvYQhKpGgNkamAI6h2lGGZq2X8ftOJvF/XjTUg==" crossorigin="anonymous" async></script>
      
    
      

      
      

      
    
      

      
      

      
    
      

      
      
        
      

      
    
      

      
      

      
    
      

      
      

      
    

  

  
  
  
    
      
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Montserrat:400,700%7CRoboto:400,400italic,700%7CRoboto+Mono&display=swap">
    
  

  
  
  
  
  <link rel="stylesheet" href="/www.moths.life/css/wowchemy.css">

  




  

  


  
  

  

  <link rel="manifest" href="/www.moths.life/en/index.webmanifest">
  <link rel="icon" type="image/png" href="/www.moths.life/images/icon_huacb3a23709378bffcbad6f45443e583f_197269_32x32_fill_lanczos_center_2.png">
  <link rel="apple-touch-icon" type="image/png" href="/www.moths.life/images/icon_huacb3a23709378bffcbad6f45443e583f_197269_192x192_fill_lanczos_center_2.png">

  <link rel="canonical" href="www.moths.life/en/project/bioinformatics/">

  
  
  
  
  
  
  
    
  
  
  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="og:site_name" content="">
  <meta property="og:url" content="www.moths.life/en/project/bioinformatics/">
  <meta property="og:title" content="无参转录组分析 | ">
  <meta property="og:description" content="无参转录组分析的小例子."><meta property="og:image" content="/www.moths.life/en/project/bioinformatics/featured.jpg">
  <meta property="twitter:image" content="/www.moths.life/en/project/bioinformatics/featured.jpg"><meta property="og:locale" content="en-us">
  
    
    
  

  


    









<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "Article",
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "www.moths.life/en/project/bioinformatics/"
  },
  "headline": "无参转录组分析",
  
  "image": [
    "/www.moths.life/en/project/bioinformatics/featured.jpg"
  ],
  
  "datePublished": "0001-01-01T00:00:00Z",
  "dateModified": "0001-01-01T00:00:00Z",
  
  "author": {
    "@type": "Person",
    "name": "Chen-Zhu Wang"
  },
  
  "publisher": {
    "@type": "Organization",
    "name": "Group of Insect Behavioral Physiology and Neurogenetics",
    "logo": {
      "@type": "ImageObject",
      "url": "/www.moths.life/images/logo_hu38f750835a438699642b4c990c18aebb_320818_192x192_fit_lanczos_2.png"
    }
  },
  "description": "无参转录组分析的小例子."
}
</script>

  

  


  


  





  <title>无参转录组分析 | </title>

</head>


<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" class="page-wrapper  ">

  
  
  
  
  
  <script src="/www.moths.life/js/wowchemy-init.js"></script>

  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search" class="form-control">
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


  <div class="page-header">
    











  


<nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
  <div class="container">

    
    <div class="d-none d-lg-inline-flex">
      <a class="navbar-brand" href="/www.moths.life/en/"><img src="/www.moths.life/images/logo_hu38f750835a438699642b4c990c18aebb_320818_0x70_resize_lanczos_2.png" alt=""></a>
    </div>
    

    
    <button type="button" class="navbar-toggler" data-toggle="collapse"
            data-target="#navbar-content" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
    <span><i class="fas fa-bars"></i></span>
    </button>
    

    
    <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
      <a class="navbar-brand" href="/www.moths.life/en/"><img src="/www.moths.life/images/logo_hu38f750835a438699642b4c990c18aebb_320818_0x70_resize_lanczos_2.png" alt=""></a>
    </div>
    

    
    
    <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

      
      <ul class="navbar-nav d-md-inline-flex">
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
            
            
            
              
            
            
          
        

        <li class="nav-item">
          <a class="nav-link " href="/www.moths.life/en/#demo"><span>Home</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
            
            
            
              
            
            
          
        

        <li class="nav-item">
          <a class="nav-link " href="/www.moths.life/en/#Research"><span>Research</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
            
            
            
              
            
            
          
        

        <li class="nav-item">
          <a class="nav-link " href="/www.moths.life/en/#posts"><span>News</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
            
            
            
              
            
            
          
        

        <li class="nav-item">
          <a class="nav-link " href="/www.moths.life/en/#projects"><span>Projects</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
            
            
            
              
            
            
          
        

        <li class="nav-item">
          <a class="nav-link " href="/www.moths.life/en/#people"><span>People</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
            
            
            
              
            
            
          
        

        <li class="nav-item">
          <a class="nav-link " href="/www.moths.life/en/#gallery"><span>Gallery</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
            
            
            
              
            
            
          
        

        <li class="nav-item">
          <a class="nav-link " href="/www.moths.life/en/#featured"><span>Publications</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
            
            
            
              
            
            
          
        

        <li class="nav-item">
          <a class="nav-link " href="/www.moths.life/en/#contact"><span>Contact</span></a>
        </li>

        
        

      

        
      </ul>
    </div>

    <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">
      
      
      <li class="nav-item">
        <a class="nav-link js-search" href="#" aria-label="Search"><i class="fas fa-search" aria-hidden="true"></i></a>
      </li>
      

      
      
      <li class="nav-item dropdown theme-dropdown">
        <a href="#" class="nav-link" data-toggle="dropdown" aria-haspopup="true" aria-label="Display preferences">
          <i class="fas fa-moon" aria-hidden="true"></i>
        </a>
        <div class="dropdown-menu">
          <a href="#" class="dropdown-item js-set-theme-light">
            <span>Light</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-dark">
            <span>Dark</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-auto">
            <span>Automatic</span>
          </a>
        </div>
      </li>
      

      

    </ul>

  </div>
</nav>


  </div>

  <div class="page-body">
    <article class="article article-project">

  




















  
  
    
  


<div class="article-container pt-3">
  <h1>无参转录组分析</h1>

  

  


<div class="article-metadata">

  
  

  
  <span class="article-date">
    
    
      
    
    Jan 1, 0001
  </span>
  

  

  

  
  
  
  
  

  
  

</div>

  













<div class="btn-links mb-3">
  
  








  












  
  <a class="btn btn-outline-primary my-1 mr-1" href="www.moths.life/en/slides/example/" target="_blank">
    Slides
  </a>
  






</div>


</div>


<div class="article-header article-container featured-image-wrapper mt-4 mb-4" style="max-width: 720px; max-height: 405px;">
  <div style="position: relative">
    <img src="/www.moths.life/en/project/bioinformatics/featured_hu3d03a01dcc18bc5be0e67db3d8d209a6_224363_720x0_resize_q90_lanczos.jpg" alt="" class="featured-image">
    <span class="article-header-caption">Photo by rawpixel on Unsplash</span>
  </div>
</div>



  <div class="article-container">

    <div class="article-style">
      <h2 id="无参转录组分析实操">无参转录组分析实操</h2>
<h3 id="author小虫子">Author：小虫子</h3>
<h3 id="参考statquest-high-throughput-sequencing">参考：StatQuest [High Throughput Sequencing]</h3>
<h4 id="background">Background</h4>
<p>(<a href="https://www.youtube.com/playlist?list=PLblh5JKOoLUJo2Q6xK4tZElbIvAACEykp">https://www.youtube.com/playlist?list=PLblh5JKOoLUJo2Q6xK4tZElbIvAACEykp</a>)
<img src="http://cdn.liguocheng.top/blog/20200409/VLerTjISO6cV.png?imageslim" alt="mark">
<img src="http://cdn.liguocheng.top/blog/20200409/GavgM6BTyVTG.png?imageslim" alt="mark">
<img src="http://cdn.liguocheng.top/blog/20200409/Yvvl0O7y0KkO.png?imageslim" alt="mark"></p>
<h4 id="library-construction">library construction</h4>
<p><img src="http://cdn.liguocheng.top/blog/20200409/oIoecekwQSBD.png?imageslim" alt="mark"></p>
<h4 id="sequencing">Sequencing</h4>
<p><img src="http://cdn.liguocheng.top/blog/20200409/WJuQvPrb5cB1.png?imageslim" alt="mark"></p>
<h4 id="mapping">Mapping</h4>
<p><img src="http://cdn.liguocheng.top/blog/20200409/WQy4R2fLLXj4.png?imageslim" alt="mark"></p>
<h4 id="reads_count">Reads_count</h4>
<p><img src="http://cdn.liguocheng.top/blog/20200409/JS4uvmiPcAIu.png?imageslim" alt="mark"></p>
<h4 id="normalization">Normalization</h4>
<h5 id="fpkm">FPKM</h5>
<p>(Reads Per Kilobase of exon model per Million mapped reads)
RPKM= total exon reads/ (mapped reads (Millions) * exon length(KB)
total exon reads：某个样本mapping到特定基因的外显子上的所有的reads。
mapped reads (Millions) :某个样本的所有reads总和。
exon length(KB)：某个基因的长度（外显子的长度的总和，以KB为单位）。
<img src="http://cdn.liguocheng.top/blog/20200409/6NR271k74ByP.png?imageslim" alt="mark"></p>
<ol>
<li>
<p>read counts 除以基因长度
<img src="http://cdn.liguocheng.top/blog/20200409/6Dc3wBoaTGdj.png?imageslim" alt="mark">
2.将每个样品的 read counts 合并，添加一行 Total Reads
<img src="http://cdn.liguocheng.top/blog/20200409/PzDyfJTd9AOW.png?imageslim" alt="mark"></p>
</li>
<li>
<p>再除以 Total，得到以下表格
<img src="http://cdn.liguocheng.top/blog/20200409/Wns1qwP6WRhs.png?imageslim" alt="mark"></p>
</li>
</ol>
<h5 id="tpm">TPM</h5>
<p>(Transcripts Per Kilobase Million)
(推荐软件，RSEM/Stringtie) 的计算公式：
TPMi=(Ni/Li)*1000000/sum(Ni/Li+……..+ Nm/Lm)
Ni：mapping到基因i上的read数； Li：基因i的外显子长度的总和</p>
<ol start="3">
<li>
<p>将每个样品的数值求和，添加一行 Total
<img src="http://cdn.liguocheng.top/blog/20200409/Pq7DvMgC2yrO.png?imageslim" alt="mark"></p>
</li>
<li>
<p>再除以 Total，得到以下表格
<img src="http://cdn.liguocheng.top/blog/20200409/IYMy7IKQpf7o.png?imageslim" alt="mark"></p>
</li>
</ol>
<h4 id="visualization">Visualization</h4>
<p><img src="http://cdn.liguocheng.top/blog/20200409/LJmvDOXSB69U.png?imageslim" alt="mark"></p>
<h4 id="differential-expression-genes">Differential expression genes</h4>
<p><img src="http://cdn.liguocheng.top/blog/20200409/spbTn4xiwanU.png?imageslim" alt="mark"></p>
<h4 id="1-数据质控与过滤">1. 数据质控与过滤</h4>
<h5 id="fastq格式说明">FASTQ格式说明：</h5>
<p>@HWUSI-EAS100R:6:73:941:1973#0/1
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTT
+HWUSI-EAS100R:6:73:941:1973#0/1
!''<em>((((</em><strong>+))%%%++)(%%%%).1</strong>*-+*''))**55CCF&raquo;&raquo;&raquo;CCCCCCC6</p>
<ul>
<li>其中第一行以@开头，后面是reads的ID以及其他信息，例如上例中 HWUSI-EAS100R代表Illumina设备名称，6代表flowcell中的第六个lane，73代表第六个lane中的第73个tile，941:1973代表该read在该tile中的x：y坐标信息；#0，若为多样本的混合作为输入样本，则该标志代表样本的编号，用来区分个样本中的reads；/1代表paired end中的前一个read。</li>
<li>第二行为read的序列。</li>
<li>第三行以“+”开头，跟随着该read的名称（一般于@后面的内容相同），但有时可以省略，但“+”一定不能省。</li>
<li>最初sanger中心用Phred quality score来衡量该read中每个碱基的质量，既-10lgP ，其中P代表该碱基被测序错误的概率，如果该碱基测序出错的概率为0.001，则Q应该为30，那么30+33=63，那么63对应的ASCii码为“？”，则在第四行中该碱基对应的质量代表值即为&quot;?&quot;。
PHRED33: Convert quality scores to Phred-33
PHRED64: Convert quality scores to Phred-64</li>
</ul>
<h5 id="trimmomatic">Trimmomatic</h5>
<p>Example:
java -jar trimmomatic-0.35.jar PE -phred33 input_forward.fq.gz input_reverse.fq.gz output_forward_paired.fq.gz output_forward_unpaired.fq.gz output_reverse_paired.fq.gz output_reverse_unpaired.fq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36</p>
<ul>
<li>ILLUMINACLIP: Cut adapter and other illumina-specific sequences from the read.</li>
<li>SLIDINGWINDOW: Perform a sliding window trimming, cutting once the average quality within the window falls below a threshold.</li>
<li>LEADING: Cut bases off the start of a read, if below a threshold quality</li>
<li>TRAILING: Cut bases off the end of a read, if below a threshold quality</li>
<li>MINLEN: Drop the read if it is below a specified length</li>
</ul>
<blockquote>
<p>This will perform the following:
Remove adapters (ILLUMINACLIP:TruSeq3-PE.fa:2:30:10)
Remove leading low quality or N bases (below quality 3) (LEADING:3)
Remove trailing low quality or N bases (below quality 3) (TRAILING:3)
Scan the read with a 4-base wide sliding window, cutting when the average quality per base drops below 15 (SLIDINGWINDOW:4:15)
Drop reads below the 36 bases long (MINLEN:36)</p>
</blockquote>
<pre><code>cat filter.sh
trimmomatic PE -phred33 \
/Storage/data003/ligc/jiangnanRAWdata/Msep_F_Ant_1.fq.gz \
/Storage/data003/ligc/jiangnanRAWdata/Msep_F_Ant_2.fq.gz \
./Msep_F_Ant_1_paired.fq.gz ./Msep_F_Ant_1_unpaired.fq.gz \
./Msep_F_Ant_2_paired.fq.gz ./Msep_F_Ant_2_unpaired.fq.gz \
ILLUMINACLIP:/Storage/data003/ligc/anaconda3/share/trimmomatic/adapters/TruSeq3-PE-2.fa:2:30:10 LEADING:5 TRAILING:5 SLIDINGWINDOW:4:5 MINLEN:25
</code></pre>
<h4 id="2-trinity-assembly">2. Trinity assembly</h4>
<h5 id="workflow">Workflow</h5>
<p><img src="http://cdn.liguocheng.top/blog/20200409/2SmB0IEVnTOB.png?imageslim" alt="mark"></p>
<h5 id="原理">原理：</h5>
<p><strong>Inchworm module</strong>：将reads打断成kmer，通过对kmer的延伸构建出contig序列。
<strong>Chrysalis module</strong>：根据contig之间(k-1)mer的重叠关系，构建出De Bruijn graph，理想情况下一张图代表一个基因，不同的路径代表不同的剪切形式。
<strong>Butterfly module</strong>：根据reads对各个路径的支持选择最优路径，打印输出。
<img src="http://cdn.liguocheng.top/blog/20200409/g5t713zvLSDB.png?imageslim" alt="mark">
<img src="http://cdn.liguocheng.top/blog/20200409/LLutSmx0TI9S.png?imageslim" alt="mark">
<img src="http://cdn.liguocheng.top/blog/20200409/HKmVery80svd.png?imageslim" alt="mark"></p>
<p><img src="http://cdn.liguocheng.top/blog/20200409/IMevvKiSlDux.png?imageslim" alt="mark"></p>
<p><img src="http://cdn.liguocheng.top/blog/20200409/QedbSkMJduuG.png?imageslim" alt="mark"></p>
<p><strong>Inchworm</strong> assembles the RNA-seq data into the unique sequences of transcripts, often generating full-length transcripts for a dominant isoform, but then reports just the unique portions of alternatively spliced transcripts.</p>
<p><strong>Chrysalis</strong> clusters the Inchworm contigs into clusters and constructs complete de Bruijn graphs for each cluster. Each cluster represents the full transcriptonal complexity for a given gene (or sets of genes that share sequences in common). Chrysalis then partitions the full read set among these disjoint graphs.</p>
<p><strong>Butterfly</strong> then processes the individual graphs in parallel, tracing the paths that reads and pairs of reads take within the graph, ultimately reporting full-length transcripts for alternatively spliced isoforms, and teasing apart transcripts that corresponds to paralogous genes.</p>
<pre><code>nohup /Storage/data003/ligc/software/trinityrnaseq-Trinity-v2.8.4/Trinity \
 --seqType fq \
 --max_memory 60 \
 --samples_file /Storage/data003/ligc/jiangnanRAWdata/sample.txt \
 --CPU 30 \
 &gt; trinity.log 2&gt; trinity_err.log &amp;
</code></pre>
<h4 id="3-transcripts-quality-assessment">3. Transcripts quality assessment</h4>
<h5 id="31-trinity_stats">3.1 Trinity_Stats</h5>
<pre><code>TrinityStats.pl ../MsepTrinity.fa &gt; Assembly_Stats.txt
</code></pre>
<p><img src="http://cdn.liguocheng.top/blog/20200409/wfGlFmakYYbm.png?imageslim" alt="mark"></p>
<blockquote>
<p>拼接的转录本数量一般在20万条以内都可以，转录本数量远多于基因数量，如果转录本数量过多可以考虑使用corset聚类。</p>
</blockquote>
<h5 id="32-buscobenchmarking-universal-single-copy-orthologs">3.2 BUSCO(Benchmarking Universal Single-Copy Orthologs)</h5>
<p>BUSCO原理:在近缘物种中，总有一些基因是保守的。那么，我们就来看这些保守的基因拼接的完整性和准确性如何。BUSCO 软件根据OrthoDB 数据库，构建了几个大的进化分支的单拷贝基因集。将转录本拼接结果与该基因集进行比较，根据比对上的比例、完整性，来评价拼接结果的准确性和完整性。</p>
<pre><code>#下载数据库(目前本版本可以自动下载数据库)
wget http://busco.ezlab.org/v2/datasets/insecta_odb9.tar.gz
#目前数据库更新到v10了
(base) [ligc@cluster busco]$ cat Run_busco.sh
#获取每个基因最长转录本
perl /pub/anaconda3/opt/trinity-2.8.4/util/misc/get_longest_isoform_seq_per_trinity_gene.pl ../../MsepTrinity.fa &gt;longest_isoform.fasta

#运行busco
busco -i longest_isoform.fasta \
        -l insecta_odb10 \
        -o busco \
        -m tran \
        --cpu 10
</code></pre>
<p><img src="http://cdn.liguocheng.top/blog/20200409/sXsOAuiQM93b.png?imageslim" alt="short summary">
可以看到有85%的转录本序列都可以比对到完整的BUSCO.</p>
<h4 id="4-annotation">4. Annotation</h4>
<p>下载数据库文件</p>
<pre><code class="language-bash">(base) [ligc@cluster Trinotate]$ cat Build_Trinotate_db.sh
#下载database: uniprot ;eggnog ;GO ;Pfam ;pfam2go
perl /Storage/data003/ligc/software/Trinotate-Trinotate-v3.1.1/admin/Build_Trinotate_Boilerplate_SQLite_db.pl Trinotate20190422
##构建blast database;diamond的速度更快
# makeblastdb -in uniprot_sprot.pep -dbtype prot
diamond makedb --db new_uniprot_sprot.pep --in new_uniprot_sprot.pep
##构建HMMER database
gunzip Pfam-A.hmm.gz
hmmpress Pfam-A.hmm
##去除空行
awk 'BEGIN {RS = &quot;&gt;&quot; ; FS = &quot;\n&quot; ; ORS = &quot;&quot;} $2 {print &quot;&gt;&quot;$0}' uniprot_sprot.pep &gt; new_uniprot_sprot.pep
</code></pre>
<p><img src="http://cdn.liguocheng.top/blog/20200409/F8DDaEVrnHXN.png?imageslim" alt="mark"></p>
<pre><code>[ligc@cluster step_1_run_transdecoder]$ cat 1_run_transdecoder.sh
#!/bin/bash
transcripts=/Storage/data003/ligc/raw_data/MsepTrinity.fa
uniprot_sprot_db=/Storage/data003/ligc/database/Trinotate/new_uniprot_sprot.pep
Pfam=/Storage/data003/ligc/database/Trinotate/Pfam-A.hmm
#1.初步筛选ORF
TransDecoder.LongOrfs -t $transcripts
##生成longest_orfs.pep，与已知蛋白数据库进行比对，提高ORF预测的准确性
#2.与已知蛋白数据库比对（可选）
diamond blastp -q /Storage/data003/ligc/Annotation/step_1_run_transdecoder/MsepTrinity.fa.transdecoder_dir/longest_orfs.pep \
             --db $uniprot_sprot_db  \
             --max-target-seqs 1 \
             --outfmt 6 \
             --evalue 1e-10 \
             --threads 20 \
             &gt; MsepTrinity.fa.transdecoder_dir/longest_orfs.pep.blastp.outfmt6
##生成longest_orfs.pep.blastp.outfmt6
hmmscan --cpu 35 \
        --domtblout /Storage/data003/ligc/Annotation/step_1_run_transdecoder/MsepTrinity.fa.transdecoder_dir/longest_orfs.pep.pfam.domtblout \
        $Pfam \
/Storage/data003/ligc/Annotation/step_1_run_transdecoder/MsepTrinity.fa.transdecoder_dir/longest_orfs.pep
##生成longest_orfs.pep.pfam.domtblout
#3.进一步筛选ORF(如果一个ORF在另一个ORF内部则保留长的ORF）
TransDecoder.Predict -t $transcripts \
            --retain_pfam_hits MsepTrinity.fa.transdecoder_dir/longest_orfs.pep.pfam.domtblout \
            --retain_blastp_hits MsepTrinity.fa.transdecoder_dir/longest_orfs.pep.blastp.outfmt6
##生成MsepTrinity.fa.transdecoder.pep；MsepTrinity.fa.transdecoder.bed；MsepTrinity.fa.transdecoder.cds；MsepTrinity.fa.transdecoder.gff3
</code></pre>
<p><img src="http://cdn.liguocheng.top/blog/20200409/x3R1W53FR5pU.png?imageslim" alt="mark"></p>
<blockquote>
<p>hmmer 软件包中的hmmscan命令用于功能注释，而hmmsearch则用于抗性基因的鉴定和转录因子的发现。</p>
</blockquote>
<p><img src="http://cdn.liguocheng.top/blog/20200409/Od23b21rRoyG.png?imageslim" alt="mark"></p>
<pre><code class="language-bash">[ligc@cluster step_2_run_blast_hmmer]$ cat step_2_run_blast_hmmer.sh
$transcripts=/Storage/data003/ligc/raw_data/MsepTrinity.fa
$proteins=/Storage/data003/ligc/Annotation/step_1_run_transdecoder/MsepTrinity.fa.transdecoder.pep
$uniprot_sprot_db=/Storage/data003/ligc/database/Trinotate/new_uniprot_sprot.pep
$Pfam=/Storage/data003/ligc/database/Trinotate/Pfam-A.hmm
#blastx -query $transcripts -db $uniprot_sprot_db -num_threads 8 -max_target_seqs 1 -outfmt 6 &gt; blastx.outfmt6
diamond blastx --query $transcripts --db $uniprot_sprot_db --threads 35 --max-target-seqs 1 --outfmt 6 &gt; blastx.outfmt6
#blastp -query $proteins -db $uniprot_sprot_db -num_threads 8 -max_target_seqs 1 -outfmt 6 &gt; blastp.outfmt6
diamond blastp --query $proteins --db $uniprot_sprot_db --threads 35 --max-target-seqs 1 --outfmt 6 &gt; blastp.outfmt6
hmmscan --cpu 35 --domtblout TrinotatePFAM.out $Pfam $proteins &gt; pfam.log
</code></pre>
<p>blast/diamond tabular 格式说明：
<img src="http://cdn.liguocheng.top/blog/20200409/25eJaTxe1ADG.png?imageslim" alt="blast">
12列对应的含义依次是：
Query id：           查询序列ID标识
Subject id：         比对上的目标序列ID标识
% identity：         序列比对的一致性百分比
alignment length：符合比对的比对区域的长度
mismatches：      比对区域的错配数
gap openings：    比对区域的gap数目
q. start：             比对区域在查询序列(Query id)上的起始位点
q. end：              比对区域在查询序列(Query id)上的终止位点
s. start：             比对区域在目标序列(Subject id)上的起始位点
s. end：              比对区域在目标序列(Subject id)上的终止位点
e-value：            比对结果的期望值
bit score：          比对结果的bit score值</p>
<blockquote>
<p><strong>uniprot id可以mapping到GO,KEGG,eggNOG数据库</strong></p>
</blockquote>
<p><img src="http://cdn.liguocheng.top/blog/20200409/c24eiroNPQv2.png?imageslim" alt="mark"></p>
<p><img src="http://cdn.liguocheng.top/blog/20200409/fiFlUrtYsDXM.png?imageslim" alt="mark"></p>
<pre><code class="language-bash">[ligc@cluster step_3_run_trinotate]$ cat step3_run_trinotate.sh
#/bin/bash
TRINOTATE_HOME=/Storage/data003/ligc/software/Trinotate-Trinotate-v3.1.1
transcripts=/Storage/data003/ligc/raw_data/MsepTrinity.fa
proteins=/Storage/data003/ligc/Annotation/step_1_run_transdecoder/MsepTrinity.fa.transdecoder.pep
gene_trans_map=/Storage/data003/ligc/raw_data/MsepTrinity.fa.gene_trans_map
Trinotate_sqlite=/Storage/data003/ligc/database/Trinotate/Trinotate20190422.sqlite
blastp_outfmt6=/Storage/data003/ligc/Annotation/step_2_run_blast_hmmer/blastp.outfmt6
blastx_outfmt6=/Storage/data003/ligc/Annotation/step_2_run_blast_hmmer/blastx.outfmt6
pfam_out=/Storage/data003/ligc/Annotation/step_2_run_blast_hmmer/TrinotatePFAM.out

#copy Trinotate.sqlite
#cp $Trinotate_sqlite ./Trinotate.sqlite

#init
$TRINOTATE_HOME/Trinotate Trinotate.sqlite init --gene_trans_map $gene_trans_map --transcript_fasta $transcripts --transdecoder_pep $proteins

# load protein hits
$TRINOTATE_HOME/Trinotate Trinotate.sqlite LOAD_swissprot_blastp $blastp_outfmt6

# load transcript hits
$TRINOTATE_HOME/Trinotate Trinotate.sqlite LOAD_swissprot_blastx $blastx_outfmt6

# load pfam hits
$TRINOTATE_HOME/Trinotate Trinotate.sqlite LOAD_pfam $pfam_out

# report
$TRINOTATE_HOME/Trinotate Trinotate.sqlite report &gt; trinotate_annotation_report.xls
</code></pre>
<h4 id="5-expression-quantification">5. Expression quantification</h4>
<h5 id="原理-1">原理：</h5>
<p>1.在不同背景下比较mRNA水平</p>
<p>同一物种，不同组织：研究基因在不同部分的表达情况
同一物种，同一组织：研究基因在不同处理下，不同条件下的表达变化
同一组织，不同物种：研究基因的进化关系
时间序列实验： 基因在不同时期的表达情况与发育的关系</p>
<p><strong>我们在进行转录组测序的时候，并不知道用于建库测序的 RNA 来自多少个细胞，总过有多少转录本。所以只能进行相对定量(scRNA-seq)。相对定量，描述的是该基因的转录本占样本中所有转录本的百分比。</strong>
<img src="http://cdn.liguocheng.top/blog/20200409/QfKFacxjQiFU.png?imageslim" alt="mark">
<img src="http://cdn.liguocheng.top/blog/20200409/RbIgRDTTQjMV.png?imageslim" alt="mark">
TPM 中的<img src="http://cdn.liguocheng.top/blog/20200409/8GEo3mN2kbnt.png?imageslim" alt="mark">是对所有基因转录本数目的求和。FPKM 的公式中，<img src="http://cdn.liguocheng.top/blog/20200409/9i66VLtXAugo.png?imageslim" alt="mark">没有考虑基因长度的影响，相当于直接使用了总 Fragments 数目。
所以，TPM 更符合我们对相对表达量的定义，而FPKM 似乎没有明确的意义。</p>
<h5 id="脚本">脚本：</h5>
<pre><code class="language-bash">(base) [ligc@cluster quantification]$ cat quantify.sh
perl /Storage/data003/ligc/software/trinityrnaseq-Trinity-v2.8.4/util/align_and_estimate_abundance.pl --transcripts /Storage/data003/ligc/raw_data/MsepTrinity.fa --seqType fq --est_method RSEM --aln_method bowtie2 --prep_reference --trinity_mode --samples_file /Storage/data003/ligc/jiangnanRAWdata/sample.txt --thread_count 40

ls */RSEM.genes.results &gt; genes.quant_files.txt
perl /Storage/data003/ligc/software/trinityrnaseq-Trinity-v2.8.4/util/abundance_estimates_to_matrix.pl --est_method RSEM --cross_sample_norm TMM --name_sample_by_basedir --gene_trans_map /Storage/data003/ligc/raw_data/MsepTrinity.fa.gene_trans_map --quant_files isoforms.quant_files.txt --out_prefix isoforms
</code></pre>
<p>align_and_estimate_abundance.pl 将原始reads比对到组装好的转录本上生成各个样本基因的相对表达量矩阵（TPM,以及FPKM值）。
abundance_estimates_to_matrix.pl 将不同样本的表达量合并为一个表达矩阵。
#####结果：
<img src="http://cdn.liguocheng.top/blog/20200409/0syhlMhDEpvz.png?imageslim" alt="mark"></p>
<pre><code>less -S RSEM.genes.results
</code></pre>
<p><img src="http://cdn.liguocheng.top/blog/20200409/U2D59RVcTLth.png?imageslim" alt="mark"></p>
<h4 id="6-differential-expression-analysis">6. Differential Expression Analysis</h4>
<blockquote>
<p>使用edgeR进行无生物学重复的差异表达分析</p>
</blockquote>
<pre><code class="language-bash">[ligc@cluster DE_analysis]$ cat run_DE_analysis.sh
#/bin/bash
TRINITY_HOME = /Storage/data003/ligc/software/trinityrnaseq-Trinity-v2.8.4
isoform_counts_matrix = /Storage/data003/ligc/quantification/genes.isoform.counts.matrix
samples_files = /Storage/data003/ligc/jiangnanRAWdata/sample.txt

perl /Storage/data003/ligc/software/trinityrnaseq-Trinity-v2.8.4/Analysis/DifferentialExpression/run_DE_analysis.pl \
    --matrix /Storage/data003/ligc/quantification/isoforms.isoform.counts.matrix \
    --method edgeR \
    --dispersion 0.1
    --samples_file /Storage/data003/ligc/jiangnanRAWdata/sample.txt
</code></pre>
<p>#####结果分析：
<img src="http://cdn.liguocheng.top/blog/20200409/uEdeTMq9Gpfv.png?imageslim" alt="mark">
<strong>logFC : log(fold change) 表示两个样本间表达量的倍数的对数。</strong>
<strong>logCPM : log(counts per millions) 可以理解为未经过基因长度标准化的表达量的对数。</strong>
<strong>PValue：</strong>
<strong>FDR:</strong>
<strong>DEG_list : FDR &lt; 0.05 &amp;&amp; | logFC | &gt;1</strong></p>
<h4 id="7normalization">7.Normalization</h4>
<p>一：样品间（组间的数据）的表达标准化
(处理样品中，有一个基因的表达量异常的高，大量的reads被占，导致其他gene的相对表达量下降。)TPM的标准化是在各自的样品内部的差异基因相对表达量。对于组间来说，也就是处理组和对照组来说就需要进一步的标准化。目的就是能够得到样品间的绝对表达量是否有差异。
标准化方法：
内参基因，看家基因。某一基因在样本中的表达肯定是一样的。(有限制，依赖基因的功能注释，数目少准确不高。
<img src="http://cdn.liguocheng.top/blog/20200409/sGJRkgrIwjlT.png?imageslim" alt="mark">
解决方法2：假设大多数基因是没有差异表达的。用统计学方法找到标准化因子。
DESeq2(estimateSizeFactors/sizeFactors) edgeR（TMM,calcNormFactors）
<em>可用Trinity软件包里的run_DE_analysis.pl</em>
二：假设检验进行差异表达基因的鉴定
1.建库测序是一个随机抽样的过程。纵使是两次测序，即抽样，同一基因的reads_count也会有差异。
当观察到的基因在两组reads_count 中不一样的时候，可能有两种原因：可能是表达丰度上的差异，也可能是随机抽样过程中的波动。假设是随机波动导致的，然后计算在随机波动的前提下，出现当前事件的概率，如果概率低。
2.假设检验：需要根据实际情况（总体均值、方差是否已知，单样本还是多样本，重复个数的多少/大样本量还是小样本量）
统计方法：z检验；t检验；卡方检验
3.差异表达基因通常用t检验（组间差异除以/组内差异，组间差异更大，而组内差异小则更可能有差异【此即是需要测更多生物学重复，能更准确的估计组内差异】）
4.t值的双侧检验和单侧检验。
单侧检验：G1在处理组中表达 大于 对照组。右侧面积-&gt;则单侧检验，即up-regulate。
双侧检验：G1处理组与对照组有明显差异。则双侧检验
5.一类错误(即为假阳性，与P-value对应，0.05 or 0.01)，二类错误(指漏诊率,假阴性)
降低一类错误：对于一条gene可能有5%误诊，而在差异基因鉴定时都是大规模的，如果一万条，50条误诊则太高了。p-value 进一步的矫正： FDR，q-value
降低二类错误：G3增加生物学重复，G2增加测序量来提高抽样次数。
<img src="http://cdn.liguocheng.top/blog/20200409/tyKz733fdCVu.png?imageslim" alt="mark"></p>
<h4 id="8gokegg-pathway-enrichment-analysis">8.GO&amp;KEGG pathway Enrichment Analysis</h4>
<p><strong>基因富集分析</strong>是在一组基因中找到具有一定基因功能特征和生物过程的基因集，在研究差异表达基因、筛选基因的后续分析中经常使用。
<strong>算法</strong>
大体上富集分析有四类算法：ORA、FCS、PT、NT
<img src="http://cdn.liguocheng.top/blog/20200409/W5G1GeLHewkd.png?imageslim" alt="algorithm">
<strong>【最常用】ORA(Over Representation Analysis)：过表达分析</strong>
<strong>GO富集分析：</strong>
<strong>推荐使用clusterProfiler软件包（R包）</strong></p>
<ul>
<li>
<p>Gene Ontology: 基因本体论，描述基因的层级关系【基于ORA算法】可以算得上是高通量数据分析的标配，转录组、甲基化、ChIP-seq、重测序等，都会用到对一个或多个集合的基因进行功能富集分析，来找这个基因集的功能偏好性。</p>
</li>
<li>
<p>它将基因分门别类放入细胞组分CC、分子功能MF和生物过程BP三个功能类别中（分别对应基因产物在哪里发挥功能、发挥什么样的功能、怎样发挥功能）。
<strong>KEGG富集分析：</strong>
Kyoto Encyclopedia of Genes and Genomes: 系统分析基因产物和化合物在细胞中的代谢途径以及这些基因产物的功能的数据库【基于ORA算法】</p>
</li>
<li>
<p>整合了基因组、化学分子和生化系统等方面的数据,包括代谢通路（KEGG PATHWAY）、药物（KEGG DRUG）、疾病（KEGG DISEASE）、功能模型（KEGG MODULE）、基因序列（KEGG GENES）及基因组（KEGG GENOME）等等。</p>
</li>
<li>
<p>KEGG 有一套完整KO注释的系统，可完成新测序物种的基因组或转录组的功能注释（KO是蛋白质或酶的一个分类体系，将同一条通路上功能相似、序列相似的蛋白质归为一类）。因此它可以将基因一个个归置到代谢网络指定位置上。</p>
<h4 id="参考文章">参考文章：</h4>
<p>1.<a href="https://www.jianshu.com/p/042b888d5520" target="_blank" rel="noopener">https://www.jianshu.com/p/042b888d5520</a>
2.<a href="https://www.jianshu.com/p/d09e624efcab" target="_blank" rel="noopener">https://www.jianshu.com/p/d09e624efcab</a>
3.<a href="https://www.jianshu.com/p/1505fa220ce4" target="_blank" rel="noopener">https://www.jianshu.com/p/1505fa220ce4</a>
4.<a href="https://www.jianshu.com/nb/14291282" target="_blank" rel="noopener">https://www.jianshu.com/nb/14291282</a>
5.<a href="http://tiramisutes.github.io/2019/01/21/Trinity.html" target="_blank" rel="noopener">http://tiramisutes.github.io/2019/01/21/Trinity.html</a>
6.<a href="https://www.jianshu.com/p/042b888d5520" target="_blank" rel="noopener">https://www.jianshu.com/p/042b888d5520</a>
7.<a href="http://wap.sciencenet.cn/blog-1113671-1038659.html?mobile=1" target="_blank" rel="noopener">http://wap.sciencenet.cn/blog-1113671-1038659.html?mobile=1</a></p>
</li>
</ul>

    </div>

    






<div class="article-tags">
  
  <a class="badge badge-light" href="www.moths.life/en/tag/bioinformatics/">Bioinformatics</a>
  
</div>



<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=www.moths.life/en/project/bioinformatics/&amp;text=%e6%97%a0%e5%8f%82%e8%bd%ac%e5%bd%95%e7%bb%84%e5%88%86%e6%9e%90" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=www.moths.life/en/project/bioinformatics/&amp;t=%e6%97%a0%e5%8f%82%e8%bd%ac%e5%bd%95%e7%bb%84%e5%88%86%e6%9e%90" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=%e6%97%a0%e5%8f%82%e8%bd%ac%e5%bd%95%e7%bb%84%e5%88%86%e6%9e%90&amp;body=www.moths.life/en/project/bioinformatics/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=www.moths.life/en/project/bioinformatics/&amp;title=%e6%97%a0%e5%8f%82%e8%bd%ac%e5%bd%95%e7%bb%84%e5%88%86%e6%9e%90" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="whatsapp://send?text=%e6%97%a0%e5%8f%82%e8%bd%ac%e5%bd%95%e7%bb%84%e5%88%86%e6%9e%90%20www.moths.life/en/project/bioinformatics/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=www.moths.life/en/project/bioinformatics/&amp;title=%e6%97%a0%e5%8f%82%e8%bd%ac%e5%bd%95%e7%bb%84%e5%88%86%e6%9e%90" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>











  
  



  
  
  
  
  
  <div class="media author-card content-widget-hr">
    
      
      <a href="www.moths.life/en/author/chen-zhu-wang/"><img class="avatar mr-3 avatar-circle" src="/www.moths.life/en/author/chen-zhu-wang/avatar_hueaba847bb9db1850c15567df33143d21_113135_270x270_fill_lanczos_center_2.png" alt="Chen-Zhu Wang"></a>
    

    <div class="media-body">
      <h5 class="card-title"><a href="www.moths.life/en/author/chen-zhu-wang/">Chen-Zhu Wang</a></h5>
      <h6 class="card-subtitle">Principal investigator</h6>
      <p class="card-text">My research interests focuses on insect-plant interactions and chemical communications of insects.</p>
      <ul class="network-icon" aria-hidden="true">
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="/www.moths.life/en/czwang@ioz.ac.cn" >
        <i class="fas fa-envelope"></i>
      </a>
    </li>
  
    
    
    
    
    
    
    
      
    
    <li>
      <a href="https://www.researchgate.net/profile/Chen-Zhu_Wang" target="_blank" rel="noopener">
        <i class="ai ai-researchgate"></i>
      </a>
    </li>
  
</ul>

    </div>
  </div>












<div class="article-widget">
  
<div class="post-nav">
  
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="www.moths.life/en/project/electrophysiology/" rel="next">雄性棉铃虫利用气味结合蛋白(OBP)感受雌性性信息素(Z11-16:Ald)</a>
  </div>
  
  
</div>

</div>





  
  





    <div class="project-related-pages content-widget-hr">
      
      

      
      
      

      
      
      

      
      
      
    </div>
  </div>
</article>
  </div>

  <div class="page-footer">
    
    
    <div class="container">
      <footer class="site-footer">
  

  <p class="powered-by">
    ©2020 <a href="www.moths.life">Chen-Zhu Wang&rsquo;s Lab</a>
  </p>

  
  





  
  
  
  

  
  
    
  
  
    
  

  

  
  <p class="powered-by copyright-license-text">
    This work is licensed under <a href="https://creativecommons.org/licenses/by-nc-nd/4.0" rel="noopener noreferrer" target="_blank">CC BY NC ND 4.0</a>
  </p>
  

  <p class="powered-by footer-license-icons">
    <a href="https://creativecommons.org/licenses/by-nc-nd/4.0" rel="noopener noreferrer" target="_blank" aria-label="Creative Commons">
      <i class="fab fa-creative-commons fa-2x" aria-hidden="true"></i>
      <i class="fab fa-creative-commons-by fa-2x" aria-hidden="true"></i>
      
        <i class="fab fa-creative-commons-nc fa-2x" aria-hidden="true"></i>
      
      
        <i class="fab fa-creative-commons-nd fa-2x" aria-hidden="true"></i>
      
    </a>
  </p>




  <p class="powered-by">
    
    
    
    Published with
    <a href="https://wowchemy.com" target="_blank" rel="noopener">Wowchemy</a>  —
    the free, <a href="https://github.com/wowchemy/wowchemy-hugo-modules" target="_blank" rel="noopener">
    open source</a> website builder that empowers creators.
    
  </p>
</footer>

    </div>
    
  </div>

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

      

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>

      
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      

      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.js" integrity="sha256-yt2kYMy0w8AbtF89WXb2P1rfjcP/HTHLT7097U8Y5b8=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/instant.page/5.1.0/instantpage.min.js" integrity="sha512-1+qUtKoh9XZW7j+6LhRMAyOrgSQKenQ4mluTR+cvxXjP1Z54RxZuzstR/H9kgPXQsVB8IW7DMDFUJpzLjvhGSQ==" crossorigin="anonymous"></script>

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/highlight.min.js" integrity="sha512-TDKKr+IvoqZnPzc3l35hdjpHD0m+b2EC2SrLEgKDRWpxf2rFCxemkgvJ5kfU48ip+Y+m2XVKyOCD85ybtlZDmw==" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/languages/r.min.js"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/languages/latex.min.js"></script>
        
      

    

    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.7.1/leaflet.min.js" integrity="sha512-SeiQaaDh73yrb56sTW/RgVdi/mMqNeM2oBwubFHagc5BkixSpP1fvqF47mKzPGWYSSy4RwbBunrJBQ4Co8fRWA==" crossorigin="anonymous"></script>
    

    
    
    <script>const code_highlighting = true;</script>
    

    
    
    
    
    
    
    <script>
      const search_config = {"indexURI":"/www.moths.life/en/index.json","minLength":1,"threshold":0.3};
      const i18n = {"no_results":"No results found","placeholder":"Search...","results":"results found"};
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks",
        'slides' : "Slides"
        };
    </script>
    

    
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    

    
    

    
    
    

    
    

    
    

    
    
    
    
    
    
    
    
    
      
    
    
    
    
    <script src="/www.moths.life/js/wowchemy.min.59419d2c9809de125013c581da371866.js"></script>

    






</body>
</html>
